File: backend/go-chi-hive/internal/cells/article/ports.go
package article

import (
	"context"

	"github.com/sakrafux/realworld-tech-comparison/backend/go-chi-hive/internal/cells/user"
)

// --- Commands & Queries ---

type CreateArticleCommand struct {
	AuthorID    int64
	Title       string
	Description string
	Body        string
	TagList     []string
}

type UpdateArticleCommand struct {
	Slug        string
	UserID      int64
	Title       *string
	Description *string
	Body        *string
}

type DeleteArticleCommand struct {
	Slug   string
	UserID int64
}

type FavoriteArticleCommand struct {
	Slug   string
	UserID int64
}

type UnfavoriteArticleCommand struct {
	Slug   string
	UserID int64
}

type GetArticleQuery struct {
	Slug       string
	ObserverID *int64
}

type GetArticlesQuery struct {
	Tag        *string
	Author     *string
	Favorited  *string
	Limit      int
	Offset     int
	ObserverID *int64
}

type GetFeedQuery struct {
	UserID int64
	Limit  int
	Offset int
}

// --- Cell Ports ---

// Service defines the inbound port for article and tag related use cases.
type Service interface {
	CreateArticle(ctx context.Context, cmd CreateArticleCommand) (*Article, error)
	GetArticle(ctx context.Context, query GetArticleQuery) (*Article, error)
	GetArticles(ctx context.Context, query GetArticlesQuery) ([]*Article, int, error)
	GetFeed(ctx context.Context, query GetFeedQuery) ([]*Article, int, error)
	UpdateArticle(ctx context.Context, cmd UpdateArticleCommand) (*Article, error)
	DeleteArticle(ctx context.Context, cmd DeleteArticleCommand) error
	FavoriteArticle(ctx context.Context, cmd FavoriteArticleCommand) (*Article, error)
	UnfavoriteArticle(ctx context.Context, cmd UnfavoriteArticleCommand) (*Article, error)

	GetTags(ctx context.Context) ([]Tag, error)
}

// Repository defines the outbound port for article and tag data persistence.
type Repository interface {
	Create(ctx context.Context, article *Article, authorID int64) error
	GetBySlug(ctx context.Context, slug string, observerID *int64) (*Article, error)
	GetByTitle(ctx context.Context, title string, observerID *int64) (*Article, error)
	GetArticles(ctx context.Context, params GetArticlesQuery) ([]*Article, int, error)
	GetFeed(ctx context.Context, userID int64, limit, offset int) ([]*Article, int, error)
	Update(ctx context.Context, article *Article) error
	Delete(ctx context.Context, id int64) error
	Favorite(ctx context.Context, articleID, userID int64) error
	Unfavorite(ctx context.Context, articleID, userID int64) error

	FindAllTags(ctx context.Context) ([]Tag, error)
}

// UserProvider defines the interface for information this cell needs from the user cell.
// It matches a subset of user.Service.
type UserProvider interface {
	GetUser(ctx context.Context, id int64) (*user.User, error)
	GetUserByUsername(ctx context.Context, username string) (*user.User, error)
}

File: backend/typescript-node-express-hive/src/cells/article/ports.ts
import type { Article, Tag } from "./article.js";
import type { User } from "../user/user.js";

export interface CreateArticleCommand {
    authorId: number;
    title: string;
    description: string;
    body: string;
    tagList: string[];
}

export interface UpdateArticleCommand {
    slug: string;
    userId: number;
    title?: string;
    description?: string;
    body?: string;
}

export interface DeleteArticleCommand {
    slug: string;
    userId: number;
}

export interface FavoriteArticleCommand {
    slug: string;
    userId: number;
}

export interface UnfavoriteArticleCommand {
    slug: string;
    userId: number;
}

export interface GetArticleQuery {
    slug: string;
    observerId?: number;
}

export interface ArticleService {
    createArticle(cmd: CreateArticleCommand): Promise<Article>;
    getArticle(query: GetArticleQuery): Promise<Article>;
    updateArticle(cmd: UpdateArticleCommand): Promise<Article>;
    deleteArticle(cmd: DeleteArticleCommand): Promise<void>;
    favoriteArticle(cmd: FavoriteArticleCommand): Promise<Article>;
    unfavoriteArticle(cmd: UnfavoriteArticleCommand): Promise<Article>;
    getTags(): Promise<Tag[]>;
}

export interface ArticleRepository {
    create(article: Article, authorId: number): Promise<void>;
    getBySlug(slug: string, observerId?: number): Promise<Article | null>;
    getByTitle(title: string, observerId?: number): Promise<Article | null>;
    update(article: Article): Promise<void>;
    delete(id: number): Promise<void>;
    favorite(articleId: number, userId: number): Promise<void>;
    unfavorite(articleId: number, userId: number): Promise<void>;
    findAllTags(): Promise<Tag[]>;
}

export interface UserProvider {
    getUser(id: number): Promise<User>;
    getUserByUsername(username: string): Promise<User>;
}
